Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -101,6 +105,29 @@ protected lemma one [DecidableEq n] : PosSemidef (1 : Matrix n n R) :=
   ⟨isHermitian_one, fun x => by
     rw [one_mulVec]; exact Fintype.sum_nonneg fun i => star_mul_self_nonneg _⟩
 
+protected theorem natCast [DecidableEq n] (d : ℕ) :
+    PosSemidef (d : Matrix n n R) :=
+  ⟨isHermitian_natCast _, fun x => by
+    simp only [natCast_mulVec, dotProduct_smul]
+    rw [← nsmul_eq_smul_cast]
+    refine nsmul_nonneg (dotProduct_star_self_nonneg _) _⟩
+
+protected theorem ofNat [DecidableEq n] (d : ℕ) [d.AtLeastTwo] :
+    PosSemidef (no_index (OfNat.ofNat d) : Matrix n n R) :=
+  .natCast d
+
+protected theorem intCast [DecidableEq n] (d : ℤ) (hd : 0 ≤ d) :
+    PosSemidef (d : Matrix n n R) :=
+  ⟨isHermitian_intCast _, fun x => by
+    simp only [intCast_mulVec, dotProduct_smul]
+    rw [← zsmul_eq_smul_cast]
+    refine zsmul_nonneg (dotProduct_star_self_nonneg _) hd⟩
+
+@[simp]
+protected theorem _root_.Matrix.posSemidef_intCast_iff [DecidableEq n] [Nonempty n] (d : ℤ) :
+    PosSemidef (d : Matrix n n R) ↔ 0 ≤ (d : R) :=
```

Comment:
Can you not make this `0 ≤ (d : ℤ)` here?

Reply:
`Int.cast_pos` doesn't apply in a `StarOrderedRing`; are there cases where it doesn't hold?

Reply:
No, it works (aside from the zero ring, obviously), but we don't have the `Int` stuff generalized. It all generalizes pretty trivially just like we did for `Nat` a while ago. We're also missing an instance for `StarOrderedRing`, namely `ZeroLEOneClass`.
```lean
import Mathlib

namespace Int

variable {R : Type*} [AddCommGroupWithOne R] [PartialOrder R] [CovariantClass R R (· + ·) (· ≤ ·)]
variable [ZeroLEOneClass R] [NeZero (1 : R)]

lemma cast_mono' : Monotone (Int.cast : ℤ → R) := by
  intro m n h
  rw [← sub_nonneg] at h
  lift n - m to ℕ using h with k hk
  rw [← sub_nonneg, ← cast_sub, ← hk, cast_natCast]
  exact k.cast_nonneg'

variable {m n : ℤ}

@[simp] lemma cast_nonneg' : ∀ {n : ℤ}, (0 : R) ≤ n ↔ 0 ≤ n
  | (n : ℕ) => by simp
  | -[n+1] => by
    have : -(n : R) < 1 := lt_of_le_of_lt (by simp) zero_lt_one
    simpa [(negSucc_lt_zero n).not_le, ← sub_eq_add_neg, le_neg] using this.not_le

@[simp, norm_cast] lemma cast_le' : (m : R) ≤ n ↔ m ≤ n := by
  rw [← sub_nonneg, ← cast_sub, cast_nonneg', sub_nonneg]

lemma cast_strictMono' : StrictMono (fun x : ℤ => (x : R)) :=
  strictMono_of_le_iff_le fun _ _ => cast_le'.symm

@[simp, norm_cast] lemma cast_lt' : (m : R) < n ↔ m < n := cast_strictMono'.lt_iff_lt

@[simp] lemma cast_nonpos' : (n : R) ≤ 0 ↔ n ≤ 0 := by rw [← cast_zero, cast_le']

@[simp] lemma cast_pos' : (0 : R) < n ↔ 0 < n := by rw [← cast_zero, cast_lt']

@[simp] lemma cast_lt_zero' : (n : R) < 0 ↔ n < 0 := by rw [← cast_zero, cast_lt']

end Int

variable {R : Type*} [Ring R] [StarRing R] [PartialOrder R] [StarOrderedRing R] [NeZero (1 : R)]

-- generalizes to `Semiring R`
instance : ZeroLEOneClass R where
  zero_le_one := by simpa using star_mul_self_nonneg (1 : R)

open Int
example {n : ℤ} : (0 : R) < n ↔ 0 < n := by rw [← cast_zero, cast_lt']
```

Reply:
Do you have time to PR the above?

Reply:
Maybe not for a few days. If you don't get to it before me, I'll do it.

Reply:
@eric-wieser see #15185 and #15186

Reply:
Thanks! I've simplified the lemma statements now.

Reply:
@eric-wieser It looks like this one is still `↔ 0 ≤ (d : R)` instead of `↔ 0 ≤ d`.

Reply:
```suggestion
    PosSemidef (d : Matrix n n R) ↔ 0 ≤ d :=
```